### PR TITLE
get_config_directory: Let user set project_dir

### PR DESCRIPTION
**Pull request by sils1297**

### CODE REVIEW

**Comment by sils1297** [comment]
hi I'm going to sleep, maybe @AbdealiJK @abhsag24 @hypothesist or @Makman2 can have a look at this tomorrow?

**Comment by sils1297** [comment]
This is needed for fixing the coala bears tests!

File: tests/coalaCITest.py [hunk 1]
```diff
@@ -105,10 +105,8 @@ def test_coala_delete_orig(self):
              make_temp(suffix='.coafile', prefix='', dir=tempdir) as coafile,\
              make_temp(dir=tempdir) as unrelated_file:
             orig_file.close()
-            with open(coafile, mode="w") as file:
-                file.write("files = something\n")
 
-            execute_coala(coala_ci.main, "coala-ci",
-                          "-c", re.escape(coafile))
+            execute_coala(coala_ci.main, "coala-ci", "-S",
+                          "project_dir="+os.path.dirname(coafile))
```
Comment (AbdealiJK) [inline tests/coalaCITest.py]: space around the plus sign
Newcomer issue please

File: coalib/settings/ConfigurationGathering.py [hunk 2]
```diff
@@ -227,6 +227,13 @@ def get_config_directory(section):
     >>> get_config_directory(section) == section['files'].origin
     True
 
+    The user can manually set a project directory with the ``project_dir``
+    setting:
+
+    >>> section.append(Setting('project_dir', os.path.abspath('/tmp'), '/')
+    >>> get_config_directory(section) == os.path.abspath('/tmp')
+    True
+
```
Comment (AbdealiJK) [inline coalib/settings/ConfigurationGathering.py]: You should probably summarize the order of priority somewhere.

 1. `config_dir` from coafile
 2. origin of `file` from coafile
 3. coafile itself, if coafile is a directory
 4. coafile's directory if it is a file

**Comment by AbdealiJK** [comment]
Very small changed mainly meant for better documentation. Marking as wip. But acking it so that you can reack later (as it's only documentation)

**Comment by AbdealiJK** [comment]
Oops, tests are failing. No acks for you

File: coalib/settings/ConfigurationGathering.py [hunk 3]
```diff
@@ -238,8 +245,11 @@ def get_config_directory(section):
     if section is None:
         return os.getcwd()
 
-    path = os.path.abspath(section.get('files', '').origin)
-    return path if os.path.isdir(path) else os.path.dirname(path)
+    if 'project_dir' in section:
+        return path(section.get('project_dir'))
+
+    config = os.path.abspath(section.get('files', '').origin)
```
Comment (AbdealiJK) [inline coalib/settings/ConfigurationGathering.py]: We never test this in the integration tests now do we ?

Comment (sils1297) [inline coalib/settings/ConfigurationGathering.py]: no

File: tests/coalaCITest.py [hunk 4]
```diff
@@ -105,10 +105,8 @@ def test_coala_delete_orig(self):
              make_temp(suffix='.coafile', prefix='', dir=tempdir) as coafile,\
              make_temp(dir=tempdir) as unrelated_file:
             orig_file.close()
-            with open(coafile, mode="w") as file:
-                file.write("files = something\n")
 
-            execute_coala(coala_ci.main, "coala-ci",
-                          "-c", re.escape(coafile))
+            execute_coala(coala_ci.main, "coala-ci", "-S",
+                          "project_dir=" + os.path.dirname(coafile))
```
Comment (AbdealiJK) [inline tests/coalaCITest.py]: Don't you need `re.escape` for windows ?

**Comment by AbdealiJK** [comment]
ack 2561f40 47bed12

File: tests/coalaCITest.py (hunk 4 above)
Comment (abhsag24) [inline tests/coalaCITest.py]: @AbdealiJK  won't the functions automatically handle that?

Comment (AbdealiJK) [inline tests/coalaCITest.py]: I wasn't sure. The deleted line has `re.escape(coafile)`. Apparently it does handle escaping (as the tests pass)

**Comment by sils1297** [comment]
reack
de1481e README.rst: Elaborate ABOUT section
45bd608 coalaCITest: Simplify test
836ab7d get_config_directory: Let user set project_dir

**Comment by Adrianzatreanu** [comment]
@ the readme one, please add https://github.com/coala-analyzer/coala/issues/1802 to close it automatically, i didn't know its related to an issue :)

**Comment by Adrianzatreanu** [comment]
unack de1481e

**Comment by Adrianzatreanu** [comment]
reack de1481e

**Comment by sils1297** [comment]
@rultor merge

**Pull request merged**